lib/worker.go

```diff
@@ -0,0 +1,434 @@
+		return NewLocalWorker(cfg), nil
+	default:
+		return nil, fmt.Errorf("unrecognized worker type: %s", cfg.Type)
+	}
+}
+
+// LocalWorker is an in-process implementation of worker
+// TODO - finish parallelism implementation
+type LocalWorker struct {
+	coord   WorkCoordinator
+	stop    chan bool
+	cfg     *WorkerConfig
+	fetcher *fetchbot.Fetcher
+	queue   *fetchbot.Queue
+}
+
+// NewLocalWorker creates a LocalWorker with crawl configuration settings
+func NewLocalWorker(cfg *WorkerConfig) *LocalWorker {
+	return &LocalWorker{
+		stop: make(chan bool),
+		cfg:  cfg,
+	}
+}
+
+// SetDelay configures the delay between requests
+func (w *LocalWorker) SetDelay(d time.Duration) {
+	// TODO!
+}
+
+// Start the local worker reporting results to the given coordinator
+func (w *LocalWorker) Start(coord WorkCoordinator) error {
+	w.coord = coord
+	cfg := w.cfg
+
+	f := fetchbot.New(newMux(coord, cfg.RecordRedirects, cfg.RecordResponseHeaders))
+	f.DisablePoliteness = !cfg.Polite
+	f.CrawlDelay = time.Duration(cfg.DelayMilli) * time.Millisecond
+	f.UserAgent = cfg.UserAgent
+	if cfg.RecordRedirects {
+		f.HttpClient = NewRecordRedirectClient(coord)
+	}
+
+	w.fetcher = f
+
+	ch, err := coord.Queue()
+	if err != nil {
+		return err
+	}
+
+	w.queue = w.fetcher.Start()
+
+	go func() {
+		for {
+			select {
+			case fr := <-ch:
+				tg, err := NewTimedGet(fr.URL)
+				if err != nil {
+					log.Error(err.Error())
+					continue
+				}
+				if err := w.queue.Send(tg); err != nil {
+					log.Error(err.Error())
+					continue
+				}
+			case <-w.stop:
+				w.queue.Close()
+				break
+			}
+		}
+	}()
+
+	return nil
+}
+
+// Stop the worker
+func (w *LocalWorker) Stop() error {
+	w.stop <- true
+	return nil
+}
+
+// newMux creates a muxer (response multiplexer) for a fetchbot
+func newMux(coord WorkCoordinator, recordRedirects, recordHeaders bool) *fetchbot.Mux {
+	// Create the muxer
+	mux := fetchbot.NewMux()
+
+	// Handle all errors the same
+	mux.HandleErrors(fetchbot.HandlerFunc(func(ctx *fetchbot.Context, res *http.Response, err error) {
+		log.Infof("[ERR] %s %s - %s", ctx.Cmd.Method(), ctx.Cmd.URL(), err.Error())
+		coord.Completed(&Resource{Error: err.Error()})
+		return
+	}))
+
+	// Handle GET requests for html responses, to parse the body and enqueue all links as HEAD requests.
+	mux.Response().Method("GET").Handler(fetchbot.HandlerFunc(
+		func(ctx *fetchbot.Context, res *http.Response, err error) {
+
+			r := &Resource{URL: ctx.Cmd.URL().String()}
+
+			// TODO - huh? why this here? figure out & make a note
+			if recordRedirects {
+				r = &Resource{URL: NormalizeURLString(res.Request.URL)}
+			}
+
+			log.Infof("[%d] %s %s", res.StatusCode, ctx.Cmd.Method(), r.URL)
+
+			var st time.Time
+			if timedCmd, ok := ctx.Cmd.(*TimedCmd); ok {
+				st = timedCmd.Started
+			}
+
+			if err := r.HandleResponse(st, res, recordHeaders); err != nil {
+				log.Debugf("error handling get response: %s - %s", ctx.Cmd.URL().String(), err.Error())
+				return
```

Comment:
I think we still have to send some completion message back to the coordinator here, right?

Reply:
Oh, I meant more that right after this, we call `coord.Completed(r)`, but we’ll never get there if `r.HandleResponse()` returns an error. Shouldn’t we be doing something like this in the error handling code before calling `return`?

```go
if err := r.HandleResponse(st, res, recordHeaders); err != nil {
    log.Debugf("error handling get response: %s - %s", ctx.Cmd.URL().String(), err.Error())
    coord.Completed(&Resource{URL: r.URL, Error: err.Error()})
    return
}
```